File: packages/shared/src/usageStatsApi.ts
/**
 * Usage Statistics API client.
 *
 * Mirrors the Rust types in `usage_stats/queries.rs` and `usage_stats/storage.rs`.
 * All commands are registered in `usage_stats_api.rs` and exposed via Tauri.
 *
 * 三包共用（web-ui / underlay-ui）：统一实现点，避免两份重复定义。
 */
import { invoke } from '@tauri-apps/api/core';

// ============ Types (1:1 mirror of Rust types) ============

export interface DaySummary {
  total_active_secs: number;
  total_afk_secs: number;
  app_count: number;
  segment_count: number;
  longest_segment_secs: number;
}

export interface TimelineItem {
  hour: number;
  exe_path: string;
  process_name: string;
  display_name: string | null;
  color: string | null;
  category_id: number | null;
  /** Base64 PNG data URL (`data:image/png;base64,...`) or null. */
  icon: string | null;
  duration_secs: number;
  is_afk: boolean;
}

export interface TrendItem {
  date: string; // "YYYY-MM-DD"
  total_active_secs: number;
  total_afk_secs: number;
  app_count: number;
}

export interface TopAppItem {
  exe_path: string;
  process_name: string;
  display_name: string | null;
  category_id: number | null;
  category_name: string | null;
  color: string | null;
  /** Base64 PNG data URL (`data:image/png;base64,...`) or null. */
  icon: string | null;
  total_secs: number;
  segment_count: number;
  percentage: number;
}

export interface HeatmapCell {
  day_of_week: number; // 0 = Sunday
  hour: number; // 0..=23
  total_secs: number;
}

export interface AppRule {
  id: number | null;
  exe_path: string;
  process_name: string;
  display_name: string | null;
  category_id: number | null;
  color: string | null;
  exclude_from_stats: boolean;
  capture_title: boolean;
  /** Base64 PNG data URL (`data:image/png;base64,...`) or null. */
  icon: string | null;
  created_at: string;
  updated_at: string;
}

export interface Category {
  id: number;
  name: string;
  color: string | null;
  sort_order: number;
}

// ============ API client ============

export const usageStatsApi = {
  /** Day summary card (today if date omitted). */
  async daySummary(date?: string): Promise<DaySummary> {
    return invoke('usage_stats_day_summary', { date: date ?? null });
  },

  /** Per-hour timeline for a single day (today if date omitted). */
  async timeline(date?: string): Promise<TimelineItem[]> {
    return invoke('usage_stats_timeline', { date: date ?? null });
  },

  /** Daily trends for an inclusive date range. */
  async trends(start: string, end: string): Promise<TrendItem[]> {
    return invoke('usage_stats_trends', { start, end });
  },

  /** Top apps by total active time in an inclusive date range. */
  async topApps(start: string, end: string, limit?: number): Promise<TopAppItem[]> {
    return invoke('usage_stats_top_apps', { start, end, limit: limit ?? null });
  },

  /** 7x24 heatmap for an inclusive date range. */
  async heatmap(start: string, end: string): Promise<HeatmapCell[]> {
    return invoke('usage_stats_heatmap', { start, end });
  },

  // ── app_rules CRUD ──────────────────────────────────────────────────

  async listAppRules(): Promise<AppRule[]> {
    return invoke('usage_stats_list_app_rules');
  },

  async updateAppRule(rule: AppRule): Promise<void> {
    return invoke('usage_stats_update_app_rule', { rule });
  },

  async deleteAppRule(exePath: string): Promise<void> {
    return invoke('usage_stats_delete_app_rule', { exePath });
  },

  // ── categories CRUD ──────────────────────────────────────────────────

  async listCategories(): Promise<Category[]> {
    return invoke('usage_stats_list_categories');
  },

  async createCategory(name: string, color?: string): Promise<number> {
    return invoke('usage_stats_create_category', { name, color: color ?? null });
  },

  async deleteCategory(id: number): Promise<void> {
    return invoke('usage_stats_delete_category', { id });
  },
};

// ============ Helpers ============

/** Format seconds as "Xh Ym" or "Ym" or "<1m". */
export function formatDuration(secs: number): string {
  if (secs < 60) return '<1m';
  const m = Math.floor(secs / 60);
  if (m < 60) return `${m}m`;
  const h = Math.floor(m / 60);
  const remM = m % 60;
  if (remM === 0) return `${h}h`;
  return `${h}h ${remM}m`;
}

/** Format seconds as "H:MM" hours (decimal hours, 1 fraction digit). */
export function formatHours(secs: number): string {
  return (secs / 3600).toFixed(1);
}

/** Return today's date as "YYYY-MM-DD" in local timezone. */
export function todayDateStr(): string {
  return localDateStr(new Date());
}

/** Format a Date as "YYYY-MM-DD" in local timezone. */
export function localDateStr(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

/** Return date N days ago as "YYYY-MM-DD". */
export function daysAgoDateStr(days: number): string {
  const d = new Date();
  d.setDate(d.getDate() - days);
  return localDateStr(d);
}

/** Human-readable label for day-of-week (0=Sunday). */
export function dayOfWeekLabel(dow: number): string {
  return ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'][dow] ?? '';
}

File: src/apps/desktop/src/api/usage_stats_api.rs
//! Tauri commands for usage statistics.
//!
//! All commands accept a `State<'_, UsageStatsStore>` (managed in `lib.rs`
//! setup). SQLite queries are blocking, so each command clones the store
//! handle (cheap — `Arc<Pool>` internally) and runs the query on a
//! `spawn_blocking` thread to avoid stalling the Tokio runtime.

use chrono::NaiveDate;
use tauri::State;

use crate::usage_stats::queries::{
    Category, DaySummary, HeatmapCell, TimelineItem, TopAppItem, TrendItem,
};
use crate::usage_stats::storage::{AppRule, UsageStatsStore};

/// Parse `YYYY-MM-DD` or fall back to today (local timezone).
fn parse_date_or_today(s: Option<String>) -> Result<NaiveDate, String> {
    match s {
        Some(s) => NaiveDate::parse_from_str(&s, "%Y-%m-%d")
            .map_err(|e| format!("invalid date '{}': {}", s, e)),
        None => Ok(chrono::Local::now().date_naive()),
    }
}

fn parse_date(s: &str, field: &str) -> Result<NaiveDate, String> {
    NaiveDate::parse_from_str(s, "%Y-%m-%d")
        .map_err(|e| format!("invalid {} date '{}': {}", field, s, e))
}

// ── Read queries ────────────────────────────────────────────────────────

#[tauri::command]
pub async fn usage_stats_day_summary(
    store: State<'_, UsageStatsStore>,
    date: Option<String>,
) -> Result<DaySummary, String> {
    let date = parse_date_or_today(date)?;
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.day_summary(date))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_timeline(
    store: State<'_, UsageStatsStore>,
    date: Option<String>,
) -> Result<Vec<TimelineItem>, String> {
    let date = parse_date_or_today(date)?;
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.timeline_for_date(date))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_trends(
    store: State<'_, UsageStatsStore>,
    start: String,
    end: String,
) -> Result<Vec<TrendItem>, String> {
    let start = parse_date(&start, "start")?;
    let end = parse_date(&end, "end")?;
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.trends_for_range(start, end))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_top_apps(
    store: State<'_, UsageStatsStore>,
    start: String,
    end: String,
    limit: Option<usize>,
) -> Result<Vec<TopAppItem>, String> {
    let start = parse_date(&start, "start")?;
    let end = parse_date(&end, "end")?;
    let limit = limit.unwrap_or(10);
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.top_apps_for_range(start, end, limit))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_heatmap(
    store: State<'_, UsageStatsStore>,
    start: String,
    end: String,
) -> Result<Vec<HeatmapCell>, String> {
    let start = parse_date(&start, "start")?;
    let end = parse_date(&end, "end")?;
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.hourly_heatmap_for_range(start, end))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

// ── app_rules CRUD ──────────────────────────────────────────────────────

#[tauri::command]
pub async fn usage_stats_list_app_rules(
    store: State<'_, UsageStatsStore>,
) -> Result<Vec<AppRule>, String> {
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.list_app_rules())
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_update_app_rule(
    store: State<'_, UsageStatsStore>,
    rule: AppRule,
) -> Result<(), String> {
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.update_app_rule(&rule))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_delete_app_rule(
    store: State<'_, UsageStatsStore>,
    exe_path: String,
) -> Result<(), String> {
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.delete_app_rule(&exe_path))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

// ── categories CRUD ─────────────────────────────────────────────────────

#[tauri::command]
pub async fn usage_stats_list_categories(
    store: State<'_, UsageStatsStore>,
) -> Result<Vec<Category>, String> {
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.list_categories())
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_create_category(
    store: State<'_, UsageStatsStore>,
    name: String,
    color: Option<String>,
) -> Result<i64, String> {
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.create_category(&name, color.as_deref()))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_delete_category(
    store: State<'_, UsageStatsStore>,
    id: i64,
) -> Result<(), String> {
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || store.delete_category(id))
        .await
        .map_err(|e| format!("join error: {}", e))?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn usage_stats_update_category(
    store: State<'_, UsageStatsStore>,
    id: i64,
    name: Option<String>,
    color: Option<Option<String>>,
) -> Result<(), String> {
    let store = store.inner().clone();
    tokio::task::spawn_blocking(move || {
        store.update_category(id, name.as_deref(), color.as_ref().map(|c| c.as_deref()))
    })
    .await
    .map_err(|e| format!("join error: {}", e))?
    .map_err(|e| e.to_string())
}
